## Load MMRF CoMMpass data

```{r process_mmrf_commpass_data}
# mmrf gene weights output filepath
gene_expr_path     <- file.path(mmrf_dir, 'rnaseq/MMRF_CoMMpass_IA13a_E74GTF_HtSeq_Gene_Counts.txt.gz')

mmrf_clin_metadata <- file.path(mmrf_dir, 'clinical_data_tables', 'MMRF_CoMMpass_IA13_PER_PATIENT.csv')
mmrf_trmt_metadata <- file.path(mmrf_dir, 'clinical_data_tables', 'MMRF_CoMMpass_IA13_STAND_ALONE_TRTRESP.csv')
mmrf_surv_metadata <- file.path(mmrf_dir, 'clinical_data_tables/MMRF_CoMMpass_IA13_STAND_ALONE_SURVIVAL.csv')

gene_stats_path    <- file.path(gene_stats_dir, 'MMRF13a_gene_stats.csv')
gene_pvals_path    <- file.path(gene_pvals_dir, 'MMRF13a_gene_pvals.csv')

fgsea_pval_path    <- file.path(fgsea_dir, 'MMRF13a_fgsea_pval_results.csv')
```

```{r load_mmrf_data, message = FALSE}
# load patient gene expression data
gene_expr_raw <- read_tsv(gene_expr_path, col_types = cols())

# for now, limit ourselves to baseline bone marrow samples so that we have a single
# sample collected at the same relative stage for each patient
mask <- substr(colnames(gene_expr_raw), 11, 14) == '1_BM'

# first entry is mask corresponds to the patient id column which we always keep
mask[1] <- TRUE

gene_expr_raw <- gene_expr_raw[, mask]

# remove patients who dropped out?..

colnames(gene_expr_raw) <- substr(colnames(gene_expr_raw), 1, 9)

# drop rna-seq samples with no corresponding sample metadata (lose about 10%
# of patients here who dropped out of the study)
#mask <- substring(colnames(gene_expr_raw), 1, 9) %in% mmrf_surv$public_id
#message(sprintf("Excluding %d/%d patient RNA-Seq samples who dropped from study, or are missing metadata", sum(!mask), ncol(gene_expr_raw)))
#gene_expr_raw <- gene_expr_raw[, mask]

# drop "ERCC" rows (92 / 57997 genes)
mask <- !startsWith(gene_expr_raw$GENE_ID, 'ERCC')
gene_expr_raw <- gene_expr_raw[mask, ]

# drop genes with zero variance
mask <- apply(as.matrix(gene_expr_raw[, -1]), 1, var) > 0

#table(mask)
# mask
# FALSE  TRUE 
#  2817 55088 

gene_expr_raw <- gene_expr_raw[mask, ]

# create a cpm-normalized version of the raw counts
gene_expr <- gene_expr_raw
mmrf_col_sums <- colSums(gene_expr[, 2:ncol(gene_expr)])
gene_expr[, 2:ncol(gene_expr)] <- sweep(gene_expr[, 2:ncol(gene_expr)],
                                                    2, mmrf_col_sums, '/') * 1E6

# convert ENSEMBL ids to gene symbols
gene_expr$gene_symbol <- grch37$symbol[match(gene_expr$GENE_ID, grch37$ensgene)]

gene_expr <- gene_expr %>%
  select(-GENE_ID) %>%
  select(gene_symbol, everything())
```

#### MMRF CoMMpass Clinical Variables

```{r load_mmrf_patient_clinical_metadata, eval = !file.exists(gene_stats_path), message = FALSE}
# load patient clinical metadata
mmrf_clin <- read_csv(mmrf_clin_metadata, col_types = cols()) %>%
  select(public_id = PUBLIC_ID, iss_stage = R_ISS, ecog, mm_status = D_PT_mmstatus)

# limit metadata to those patients for which gene expression data is present and match
# orders
mmrf_clin <- mmrf_clin[match(colnames(gene_expr_raw)[-1], mmrf_clin$public_id), ]

#all(mmrf_clin$public_id == colnames(gene_expr_raw)[-1])
# [1] TRUE

# initialize result variable; creates a dataframe with only the gene identifiers
#gene_weights <- data.frame(ensgene = gene_expr$gene_symbol)
#colnames(gene_expr)[1] <- 'ensgene'

gene_pvals    <- gene_expr[, 1, drop = FALSE]

output_cols <- c()

for (covariate in c('iss_stage', 'ecog', 'mm_status')) {
  message(sprintf("Processing MMRF-IA13 / %s", covariate))

  # get covariate column
  response <- mmrf_clin[, covariate, drop = TRUE]

  # column name to use in output
  output_colname <- make.names(sprintf("MMRF13a_%s", covariate))
  output_cols <- c(output_cols, output_colname)

  # for factor columns, compute R^2 for a logit predicting the factor level
  if (is.character(response) || length(unique(response)) < 10) {
    response <- factor(response)

    # multinomial logistic regression
    null <- glm(response ~ 1, family = "binomial")

    # create paralellization handler
    cl <- makeCluster(MAX_THREADS, outfile='')
    registerDoParallel(cl)

    clusterExport(cl, c('response', 'null'))

    #res_genes <- parApply(cl, as.matrix(gene_expr[, -1]), 1, function(x) {
    #for (i in 1:nrow(gene_expr)) {
    dataset_gene_pvals <- parApply(cl, as.matrix(gene_expr[, -1]), 1, function(x) {
      mod <- glm(response ~ x, family = "binomial")

      GLM_PVAL_ROW_IND <- 2
      GLM_PVAL_COL_IND <- 4

      # in some cases (e.g. when x contains mostly 0's), fit will only include an
      # intercept term, making it necessary to check the coef dimensions
      coefs <- coef(summary(mod))

      if (all(dim(coefs) == c(2, 4))) {
        coefs[GLM_PVAL_ROW_IND, GLM_PVAL_COL_IND]
      } else {
        1
      }
    })

    stopCluster(cl)

    # append to result dataframes and update column names
    gene_pvals    <- cbind(gene_pvals, dataset_gene_pvals)

    # column name to use in output
    pval_cname <- make.names(sprintf("MMRF13a_%s_pval", covariate))

    colnames(gene_pvals)[ncol(gene_pvals)]       <- pval_cname
  }
}
```

#### MMRF CoMMpass Treatment Variables

```{r load_mmrf_patient_treatment_metadata, eval = !file.exists(gene_stats_path), message = FALSE}
# load patient treatment metadata
mmrf_treatment <- read_csv(mmrf_trmt_metadata, col_types = cols())

# limit metadata to those patients for which gene express data is present and match
# orders
mmrf_treatment <- mmrf_treatment[match(colnames(gene_expr)[-1], mmrf_treatment$public_id), ]

# all(mmrf_treatment$public_id == colnames(gene_expr)[-1])
# [1] TRUE

# most common treatments

# head(sort(table(mmrf_treatment$trtname), TRUE))
# 
#      Bortezomib/Lenalidomide/Dexamethasone  Bortezomib/Cyclophosphamide/Dexamethasone 
#                                        243                                        138 
#     Carfilzomib/Lenalidomide/Dexamethasone                   Bortezomib/Dexamethasone 
#                                        102                                         73 
#                 Lenalidomide/Dexamethasone Carfilzomib/Cyclophosphamide/Dexamethasone 
#                                         52                                         43 
# 

#treatments <- head(names(sort(table(mmrf_treatment$trtshnm), TRUE)))
#treatments
# [1] "Bor-Len-Dex" "Bor-Cyc-Dex" "Car-Len-Dex" "Bor-Dex"     "Len-Dex"     "Car-Cyc-Dex"

# treatment covariates associated with functional enrichment
treatments <- c('Bor-Len-Dex', 'Bor-Cyc-Dex', 'Bor-Dex', 'Car-Cyc-Dex')

# for each common treatment, build a model predict treatment first response;
for (treatment in treatments) {
  mmrf_treatment_subset <- mmrf_treatment %>%
    filter(trtshnm == treatment) %>%
    select(public_id, first_response = frespcd)

  # in cases where multiple entries for the same id / treatment are present, use the
  # first entry
  mmrf_treatment_subset <- mmrf_treatment_subset %>%
    group_by(public_id) %>%
    slice(1)

  # get rna-seq for samples corresponding to treatment
  gene_expr_mat <- as.matrix(gene_expr[, -1])
  gene_expr_mat <- gene_expr_mat[, mmrf_treatment_subset$public_id]

  # best response also explored, but r^2 for first response generally found to be
  # better.. 
  covariate <- 'first_response'

  message(sprintf("Processing MMRF-IA13 / %s / %s", treatment, covariate))

  # get covariate column
  response <- mmrf_treatment_subset[, covariate, drop = TRUE]

  # column name to use in output
  output_colname <- make.names(sprintf("mmrf_%s_%s", gsub('-', '_', treatment), covariate))
  output_cols <- c(output_cols, output_colname)

  # for factor columns, compute R^2 for a logit predicting the factor level
  if (is.character(response) || length(unique(response)) < 10) {
    response <- factor(response)

    # multinomial logistic regression
    null <- glm(response ~ 1, family = "binomial")

    # create paralellization handler
    cl <- makeCluster(MAX_THREADS, outfile='')
    registerDoParallel(cl)

    GLM_PVAL_ROW_IND <- 2
    GLM_PVAL_COL_IND <- 4

    clusterExport(cl, c('response', 'null', 'GLM_PVAL_ROW_IND', 'GLM_PVAL_COL_IND'))

    dataset_gene_pvals <- parApply(cl, gene_expr_mat, 1, function(x) {
      mod <- glm(response ~ x, family = "binomial")

      # in some cases (e.g. when x contains mostly 0's), fit will only include an
      # intercept term, making it necessary to check the coef dimensions
      coefs <- coef(summary(mod))

      if (all(dim(coefs) == c(2, 4))) {
        coefs[GLM_PVAL_ROW_IND, GLM_PVAL_COL_IND]
      } else {
        1
      }
    })

    stopCluster(cl)

    # save results
    gene_pvals    <- cbind(gene_pvals, dataset_gene_pvals)
    colnames(gene_pvals)[ncol(gene_pvals)] <- make.names(sprintf('MMRF13a_%s_pval', treatment))
  }
}
```

#### MMRF CoMMpass Survival Variables

```{r load_mmrf_patient_survival_metadata, eval = !file.exists(gene_stats_path), message = FALSE}
# load patient survival metadata
# regular and censored overall & progression-free survival
mmrf_surv <- read_csv(mmrf_surv_metadata, col_types = cols())

mmrf_surv <- mmrf_surv %>%
  select(public_id, pfsdy, pfscdy, oscdy, censos, censpfs)

# limit metadata to those patients for which gene express data is present and match
# orders
mmrf_surv <- mmrf_surv[match(colnames(gene_expr_raw)[-1], mmrf_surv$public_id), ]

#all(mmrf_surv$public_id == colnames(gene_expr_raw)[-1])
# [1] TRUE

# Kaplan-Meier survival regression
#fit <- survival::survfit(survival::Surv(pfscdy, censpfs) ~ 1, data = mmrf_surv)
#fit <- survival::survfit(survival::Surv(oscdy, censos) ~ 1, data = mmrf_surv)
#fit <- survival::survfit(survival::Surv(pfscdy, censpfs) ~ 1, data = mmrf_surv)

# create paralellization handler
cl <- makeCluster(MAX_THREADS, outfile='')
registerDoParallel(cl)

# Cox regression (survival ~ gene)
clusterExport(cl, c('mmrf_surv'))
tmp_ <- clusterCall(cl, function() library(survival))

# progression-free survival
gene_fits <- parApply(cl, as.matrix(gene_expr[, -1]), 1, function(x) {
  summary(survival::coxph(survival::Surv(mmrf_surv$pfscdy, mmrf_surv$censpfs) ~ x))
})

# Cox regression Wald P-value
gene_wald_pvals <- as.numeric(unlist(lapply(gene_fits, function(x) {
  x$waldtest['pvalue']
})))

# In a few cases (~20/55000), NA values may be generated in the above; for convenience,
# we set these to "1"
gene_wald_pvals[is.na(gene_wald_pvals)] <- 1
gene_pvals <- cbind(gene_pvals, MMRF13a_pfs_cox_pval = gene_wald_pvals)

# overall survival
gene_fits <- parApply(cl, as.matrix(gene_expr[, -1]), 1, function(x) {
  summary(survival::coxph(survival::Surv(mmrf_surv$oscdy, mmrf_surv$censos) ~ x))
})

stopCluster(cl)

# Cox regression Wald P-value
gene_wald_pvals <- as.numeric(unlist(lapply(gene_fits, function(x) {
  x$waldtest['pvalue']
})))

#sum(is.na(gene_wald_pvals))
# [1] 6

# replace missing values
gene_wald_pvals[is.na(gene_wald_pvals)] <- 1
gene_pvals    <- cbind(gene_pvals, MMRF13a_os_cox_pval = gene_wald_pvals)

# compute gene-, probe-wise distribution statistics
gene_stats <- data.frame(
  'gene_symbol'      = gene_expr$gene_symbol,
  'MMRF13a_mean'     = apply(as.matrix(gene_expr[, -1]), 1, mean),
  'MMRF13a_median'   = apply(as.matrix(gene_expr[, -1]), 1, median),
  'MMRF13a_sd'       = apply(as.matrix(gene_expr[, -1]), 1, sd),
  'MMRF13a_var'      = apply(as.matrix(gene_expr[, -1]), 1, var),
  'MMRF13a_mad'      = apply(as.matrix(gene_expr[, -1]), 1, mad),
  'MMRF13a_kurtosis' = apply(as.matrix(gene_expr[, -1]), 1, kurtosis)
)

gene_dip <- apply(as.matrix(gene_expr[, -1]), 1, dip.test)

dip_pvals <- unlist(lapply(gene_dip, function(x) { x$p.val }))
gene_stats$MMRF13a_dip_score <- -log10(pmax(MIN_PVAL, dip_pvals))

#
# next, collapse multi-mapped genes using the follow rules:
#
# p-values: keep min p-value
# stats   : keep average statistic
#

# p-values (use min)
gene_pvals <- gene_pvals %>%
  group_by(gene_symbol) %>%
  summarise_all(min)

# expression statistics
gene_stats <- gene_stats %>%
  group_by(gene_symbol) %>%
  summarise_all(mean)

# save results

write_csv(gene_stats, gene_stats_path)
write_csv(gene_pvals, gene_pvals_path) 
```

```{r mmrf_functional_enrichment, warnings = FALSE, eval = !file.exists(gene_stats_path)}
# perform functional enrichment on gene-level results
if (!file.exists(fgsea_pval_path)) {
  fgsea_res <- NULL

  message(sprintf("Measuring functional enrichment for MMRF IA13a... (pvalues)"))

  # measure enrichment for covariate p-values
  for (covariate in colnames(gene_pvals)[-1]) {
    message(sprintf("Measuring functional enrichment for %s / MMRF IA13a", covariate)) 

    for (gene_set in names(gene_sets)) {
      set.seed(1)

      stats <- setNames(gene_pvals %>% pull(covariate), gene_pvals$gene_symbol)
      
      # measure functional enrichment
      res <- fgsea(gene_sets[[gene_set]], stats, nperm = FGSEA_NPERM, nproc = 1) %>%
        select(-leadingEdge) %>%
        arrange(padj)

      fgsea_res <- rbind(fgsea_res, cbind(covariate, gene_set, res))
    }
  }	

  # store enrichment results
  message(sprintf("Saving %s", fgsea_pval_path))
  write_csv(fgsea_res, fgsea_pval_path)
}

# clean up
rm(gene_expr, gene_expr_raw)
suppressMessages(gc())
```

